src/main/java/com/linecorp/armeria/common/ServiceInvocationContext.java

```diff
@@ -149,6 +154,96 @@ public final EventLoop eventLoop() {
+     * accept an {@link Executor} to run callbacks on.
+     */
+    public final EventExecutor contextAwareEventExecutor() {
+        return new ServiceInvocationContextAwareEventExecutor(this, eventLoop());
+    }
+
+    public final Executor makeContextAware(Executor executor) {
+        return runnable -> executor.execute(makeContextAware(runnable));
+    }
+
+    public final <T> Callable<T> makeContextAware(Callable<T> callable) {
+        ServiceInvocationContext propagatedContext = this;
+        return () -> {
+            boolean mustResetContext = propagateContextIfNotPresent(propagatedContext);
+            try {
+                return callable.call();
+            } finally {
+                if (mustResetContext) {
+                    removeCurrent();
+                }
+            }
+        };
+    }
+
+    public final Runnable makeContextAware(Runnable runnable) {
+        ServiceInvocationContext propagatedContext = this;
+        return () -> {
+            boolean mustResetContext = propagateContextIfNotPresent(propagatedContext);
+            try {
+                runnable.run();
+            } finally {
+                if (mustResetContext) {
+                    removeCurrent();
+                }
+            }
+        };
+    }
+
+    public final <T extends Future<?>> GenericFutureListener<T> makeContextAware(
+            GenericFutureListener<T> listener) {
+        ServiceInvocationContext propagatedContext = this;
+        return new GenericFutureListener<T>() {
+            @Override
+            public void operationComplete(T future) throws Exception {
+                boolean mustResetContext = propagateContextIfNotPresent(propagatedContext);
+                try {
+                    listener.operationComplete(future);
+                } finally {
+                    if (mustResetContext) {
+                        removeCurrent();
+                    }
+                }
+            }
+
+            // Delegate equality checks to the original listener so removeListener works.
+            @Override
+            public boolean equals(Object obj) {
+                return listener.equals(obj);
+            }
+
+            @Override
+            public int hashCode() {
+                return listener.hashCode();
+            }
+        };
+    }
+
+    private boolean propagateContextIfNotPresent(ServiceInvocationContext propagatedContext) {
+        return mapCurrent(currentContext -> {
+            if (!currentContext.equals(propagatedContext)) {
+                throw new IllegalStateException(
+                        "Trying to call context-aware callback with context " + propagatedContext +
+                        ", but context is currently set to " + currentContext + ". This means the " +
+                        "callback was passed from one invocation to another which is not allowed. Make " +
+                        "sure you are not saving callbacks into shared state.");
```

Comment:
callback -> object created by makeContextAware() ? How about rephrasing this so it mentions makeContextAware() and contextAwareEventExecutor()?